frontend/src/pages/index.js

```diff
@@ -17,59 +17,69 @@ import {
   Row,
   Badge
 } from 'reactstrap'
+import { connect } from 'react-redux'
+import { fetchCandidates, addFilter, removeFilter } from '../actions'
+import { bindActionCreators } from 'redux'
+import { yearsenum, statusenum, rolesenum } from '../utils/enums'
 
 type Props = {
-  result: {}
+  candidates: Array<any>,
+  loading: boolean,
+  error: boolean,
+  filters: Object,
+  sort: Object
+}
+
+const mapDispatchToProps = dispatch => {
+  return bindActionCreators(
+    {
+      fetchCandidates,
+      addFilter,
+      removeFilter
+    },
+    dispatch
+  )
+}
+
+const mapStateToProps = state => {
+  return {
+    candidates: state.candidateListPage.candidates,
+    loading: state.candidateListPage.candidatesLoading,
+    error: state.candidateListPage.candidatesError,
+    filters: state.candidateListPage.filters,
+    sort: state.candidateListPage.sort
+  }
 }
 
 class HomePage extends Component<Props> {
   constructor(props) {
     super(props)
     this.state = {
-      candidates: this.props.result,
-      sortDropdownOpen: false,
-      showDropdownOpen: false,
-      showing: 'everyone'
+      candidates: this.props.candidates,
+      error: this.props.error,
+      loading: this.props.loading,
+      filters: this.props.filters,
+      sort: this.props.sort
     }
   }
 
-  toggleSort = () => {
-    this.setState(prevState => ({
-      sortDropdownOpen: !prevState.sortDropdownOpen
-    }))
-  }
-
-  toggleShow = () => {
-    this.setState(prevState => ({
-      showDropdownOpen: !prevState.showDropdownOpen
-    }))
-  }
-
-  handleClickShow = async e => {
-    const status = e.target.value
-    const response = await getCandidatesByStatus(status)
-    this.setState({
-      candidates: response.result,
-      showing: status
-    })
-  }
-
-  static async getInitialProps({ query }) {
-    // check whether query.id is real candidate
-    try {
-      const { result } = await getCandidatesByStatus('everyone')
-      if (result === undefined) {
-        return { error: 'Bad Request' }
-      }
-      return { result }
-    } catch (err) {
-      console.error('Candidate Page error: ', err.message)
-      return { error: 'Bad Request' }
-    }
+  componentDidMount() {
+    this.props.fetchCandidates()
   }
 
   render() {
-    if (this.props.error) {
+    let { candidates, error, loading, filters, sort } = this.props
+    const statusFilter = this.props.filters.statuses
+    const roleFilter = this.props.filters.roles
+    const yearFilter = this.props.filters.years
+    candidates = candidates.filter(candidate => {
+      return (
+        statusFilter.includes(candidate.status) &&
+        roleFilter.includes(candidate.role) &&
+        yearFilter.includes(candidate.year)
+      )
+    })
+    if (error) {
       return <div>Bad Fetch. Try again</div>
     }
```

Comment:
this should be handled right after you get `error` instead of processing some other stuff.